fontc/src/lib.rs

```diff
@@ -2289,4 +2290,89 @@ mod tests {
         // a delta of 70; so at normalized 0.5 (wght=550), its delta will *not* be 60
         assert_eq!(hvar.width_delta("A", &[NormalizedCoord::new(0.5)]), 83.0);
     }
+
+    fn anchor_coords(at: AnchorTable) -> (i32, i32) {
+        match at {
+            AnchorTable::Format1(at) => (at.x_coordinate() as i32, at.y_coordinate() as i32),
+            AnchorTable::Format2(at) => (at.x_coordinate() as i32, at.y_coordinate() as i32),
+            AnchorTable::Format3(at) => (at.x_coordinate() as i32, at.y_coordinate() as i32),
+        }
+    }
+
+    #[test]
+    fn compile_basic_gpos_mark_base() {
+        let temp_dir = tempdir().unwrap();
+        let build_dir = temp_dir.path();
+        let result = compile(Args::for_test(build_dir, "glyphs3/WghtVar_Anchors.glyphs"));
+
+        let font_file = build_dir.join("font.ttf");
+        assert!(font_file.exists());
+        let buf = fs::read(font_file).unwrap();
+        let font = FontRef::new(&buf).unwrap();
+
+        let gpos = font.gpos().unwrap();
+
+        let base_gid = GlyphId::new(result.get_glyph_index("A") as u16);
+        let macroncomb_gid = GlyphId::new(result.get_glyph_index("macroncomb") as u16);
+        let brevecomb_gid = GlyphId::new(result.get_glyph_index("brevecomb") as u16);
+
+        // If only we had more indirections
```

Comment:
*apologies that this turned into a largely unrelated essay*

so the types we use to represent lookups are annoying; it was one of the trickier parts of the whole scheme. They also haven't been exercised much, since we don't actually have (say) a shaper.

(One of the worst parts of this is the extensions, which have a different type and an *additional* level of indirection)

In any case, I definitely think there's room to add some methods here that would make this more convenient, and I'm trying to think what exactly would be best given the underlying structure.

I'm going to write up my thoughts here, and then i can convert it into an issue.

Ultimately all we really care about is:
- getting the lookupflag
- getting the markFilteringSet, if present
- getting access to the subtables
- being able unify extension and non-extension lookups (e.g. return the same thing for a pairpos lookup as an extension lookup with pairpos subtables)

Our current structure:

- a generic `Lookup<T>` type, where `T` is the type of the subtables
- a pair of generic `ExtensionLookup<T>` types (one each for GSUB/GPOS) where `T` is again the type of the subtable
- a pair of enums `PositionLookup` and `SubstitutionLookup` with a variant for each lookup type, with each variant wrapping a `Lookup<T>` where `T` is the subtable type for that variant.
- another pair of enums for each extension type


Let's imagine we want to simplify this so that, in the general case, the user only needs to think about the top-level `PositionLookup` and `SubstitutionLookup` types.

We can have lookupFlag & markFilteringSet fields on this enum.
Then to make accessing the subtables easy, I can current envision two versions of the API:

- we could just have one method for each format, e.g.

```rust
impl<'a> PositionLookup<'a> {
    single_pos_subtables(&self) -> Option<ArrayOfOffsets<'a, SinglePos<'a>, Offset16>> { .. }
}
```

(except this won't exactly work, since extension and non-extensions have a different offset type, which means we don't have a unified return type here; we'd need to introduce an enum to hide those two possibilities)

- *or* we could have *another* enum, with a variant for each non-extension lookup, that would look something like,

```rust
impl<'a> PositionLookup<'a> {
    fn subtables(&self) -> PositionSubtables<'a> { .. }
}

enum PositionSubtables<'a> {
    SinglePos(SomethingThatActsLikeAnArray<SinglePos<'a>>),
    PairPos(SomethingThatActsLikeAnArray<PairPos<'a>>),
    ...
}
```

The idea here is to basically change the level of granularity of the generics from the top-level lookup object down to the subtables themselves, at least from the perspective of the API.


How this would relate to the current code is unclear. It would be worth at least considering what this would look like as a systematic change, but it might be more pragmatic to treat it as syntactic sugar, on top of the existing types.

Reply:
Perhaps this should become a fontations issue?

It's largely not that bad in that 1) it matches the table structure and 2) it's typed ... EXCEPT the FontData you have to pass in all of a sudden.

Reply:
It appears this si now https://github.com/googlefonts/fontations/issues/683